api/pkg/transformer/feast/feature_retriever_test.go

```diff
@@ -2531,3 +2538,130 @@ func TestFeatureRetriever_RetriesRetrieveFeatures_MaxConcurrent(t *testing.T) {
+	feastClients[spec.ServingSource_REDIS] = mockFeast
+
+	for i := 0; i < requestEntityLen; i++ {
+		response := &feast.OnlineFeaturesResponse{
+			RawResponse: &serving.GetOnlineFeaturesResponseV2{
+				Metadata: &serving.GetOnlineFeaturesResponseMetadata{
+					FieldNames: &serving.FieldList{
+						Val: []string{
+							"driver_id",
+							"driver_trips:average_daily_rides",
+						},
+					},
+				},
+				Results: []*serving.GetOnlineFeaturesResponseV2_FieldVector{
+					{
+						Values: []*feastTypes.Value{
+							feast.StrVal(fmt.Sprintf("100%d", i+1)),
+							feast.Int32Val(1001),
+						},
+						Statuses: []serving.FieldStatus{
+							serving.FieldStatus_PRESENT,
+							serving.FieldStatus_PRESENT,
+						},
+					},
+				},
+			},
+		}
+
+		mockFeast.On("GetOnlineFeatures", mock.Anything, mock.MatchedBy(func(req *feast.OnlineFeaturesRequest) bool {
+			time.Sleep(500 * time.Millisecond)
+			return req.Project == "default"
+		})).Return(response, nil)
+	}
+
+	featureTableSpecs := []*spec.FeatureTable{
+		{
+			Project: "default",
+			Source:  spec.ServingSource_REDIS,
+			Entities: []*spec.Entity{
+				{
+					Name:      "driver_id",
+					ValueType: "STRING",
+					Extractor: &spec.Entity_JsonPath{
+						JsonPath: "$.drivers[*].id",
+					},
+				},
+			},
+			Features: []*spec.Feature{
+				{
+					Name:         "driver_trips:average_daily_rides",
+					DefaultValue: "0.0",
+					ValueType:    "DOUBLE",
+				},
+			},
+		},
+	}
+	compiledJSONPaths, err := CompileJSONPaths(featureTableSpecs, jsonpath.Map)
+	if err != nil {
+		panic(err)
+	}
+
+	compiledExpressions, err := CompileExpressions(featureTableSpecs, symbol.NewRegistry())
+	if err != nil {
+		panic(err)
+	}
+
+	jsonPathStorage := jsonpath.NewStorage()
+	jsonPathStorage.AddAll(compiledJSONPaths)
+	expressionStorage := expression.NewStorage()
+	expressionStorage.AddAll(compiledExpressions)
+	entityExtractor := NewEntityExtractor(jsonPathStorage, expressionStorage)
+
+	fr := NewFeastRetriever(feastClients,
+		entityExtractor,
+		featureTableSpecs,
+		&Options{
+			StatusMonitoringEnabled:          true,
+			ValueMonitoringEnabled:           true,
+			BatchSize:                        1,
+			FeastTimeout:                     100 * time.Millisecond,
+			FeastClientHystrixCommandName:    "TestFeatureRetriever_RetrieveFeatureOfEntityInRequest_FeastTimeout",
+			FeastClientMaxConcurrentRequests: 100,
+			CacheEnabled:                     true,
+			CacheSizeInMB:                    100,
+			CacheTTL:                         10 * time.Minute,
+		},
+		logger,
+	)
+
+	got, err := fr.RetrieveFeatureOfEntityInRequest(context.Background(), requestJson)
+	assert.NotNil(t, err)
+	assert.EqualError(t, err, "hystrix: timeout")
+
+	want := []*transTypes.FeatureTable{}
+	assert.ElementsMatch(t, got, want)
+
+	mockFeast.AssertExpectations(t)
+
+	time.Sleep(2 * time.Second)
+	assert.Less(t, runtime.NumGoroutine(), initialGoroutineCount+requestEntityLen, "goroutine leak")
```

Comment:
is this test reliable? is there possibility num of goroutine increasing due to other test?